sdk/samples/iot/sample_pnp.c

```diff
@@ -0,0 +1,428 @@
+  AZ_RETURN_IF_FAILED(az_json_writer_append_begin_object(&json_writer));
+  if (az_span_ptr(component_name) != NULL)
+  {
+    AZ_RETURN_IF_FAILED(az_json_writer_append_property_name(&json_writer, component_name));
+    AZ_RETURN_IF_FAILED(az_json_writer_append_begin_object(&json_writer));
+    AZ_RETURN_IF_FAILED(
+        az_json_writer_append_property_name(&json_writer, component_specifier_name));
+    AZ_RETURN_IF_FAILED(az_json_writer_append_string(&json_writer, component_specifier_value));
+  }
+
+  AZ_RETURN_IF_FAILED(az_json_writer_append_property_name(&json_writer, property_name));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_begin_object(&json_writer));
+  AZ_RETURN_IF_FAILED(
+      az_json_writer_append_property_name(&json_writer, desired_temp_response_value_name));
+  AZ_RETURN_IF_FAILED(append_callback(&json_writer, context));
+  AZ_RETURN_IF_FAILED(
+      az_json_writer_append_property_name(&json_writer, desired_temp_ack_code_name));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_int32(&json_writer, ack_code));
+  AZ_RETURN_IF_FAILED(
+      az_json_writer_append_property_name(&json_writer, desired_temp_ack_version_name));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_int32(&json_writer, ack_version));
+
+  if (az_span_ptr(ack_description) != NULL)
+  {
+    AZ_RETURN_IF_FAILED(
+        az_json_writer_append_property_name(&json_writer, desired_temp_ack_description_name));
+    AZ_RETURN_IF_FAILED(az_json_writer_append_string(&json_writer, ack_description));
+  }
+
+  AZ_RETURN_IF_FAILED(az_json_writer_append_end_object(&json_writer));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_end_object(&json_writer));
+
+  if (az_span_ptr(component_name) != NULL)
+  {
+    AZ_RETURN_IF_FAILED(az_json_writer_append_end_object(&json_writer));
+  }
+
+  *out_span = az_json_writer_get_bytes_used_in_destination(&json_writer);
+
+  return AZ_OK;
+}
+
+// Process the twin properties and invoke user callback for each property
+az_result pnp_process_twin_data(
+    az_json_reader* json_reader,
+    bool is_partial,
+    const az_span** sample_components_ptr,
+    int32_t sample_components_num,
+    char* scratch_buf,
+    int32_t scratch_buf_len,
+    pnp_property_callback property_callback,
+    void* context_ptr)
+{
+  az_json_reader copy_json_reader;
+  int32_t version;
+  int32_t len;
+  int32_t index;
+
+  AZ_RETURN_IF_FAILED(az_json_reader_next_token(json_reader));
+
+  if (!is_partial
+      && (sample_json_child_token_move(json_reader, sample_iot_hub_twin_desired) != AZ_OK))
+  {
+    printf("Failed to get desired property\r\n");
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  copy_json_reader = *json_reader;
+  if ((sample_json_child_token_move(&copy_json_reader, sample_iot_hub_twin_desired_version)
+       != AZ_OK)
+      || az_failed(az_json_token_get_int32(&(copy_json_reader.token), (int32_t*)&version)))
+  {
+    printf("Failed to get version\r\n");
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
```

Comment:
This is interesting. We are theoretically doing two passes over the data. I wonder if this can be written differently, but don't fully understand what the intended goal is yet.

Reply:
Here goal is to read the version of the document and then pass this version to the callback with each property in the document.
e.g:

{
   "thermostat1": {
    "__t": "c",
    "targetTemperature": 6662
   },
   "thermostat2": {
    "__t": "c",
    "targetTemperature": 20022
   },
   "$version": 69
  }

Now we want to call user callback with following calls
callback(thermostat1, targetTemperature, 6662, 69)
callback(thermostat2, targetTemperature, 20022, 69)



---
In reply to: [465978572](https://github.com/Azure/azure-sdk-for-c/pull/978#discussion_r465978572) [](ancestors = 465978572)

Reply:
I believe the problem is the inability to go backwards. Some data has different meanings depending on what follows it. So we have to look ahead to determine what to do with it.